### ISSUE TITLE
GapCursor doesn't work with selection false nodes

### ISSUE THREAD

**Issue by wes-r** [ProseMirror/prosemirror#844]
### Issue details

GapCursor does not properly handle `isLeaf` nodes. I can see that [gapcursor.js#L63](https://github.com/ProseMirror/prosemirror-gapcursor/blob/master/src/gapcursor.js#L63) is getting the first or last child which is always `null` for leafs.

### Steps to reproduce

Create a leaf node (`isLeaf = true`) at the top of the document. Set cursor below this leaf node and try to arrow up above it. The expectation is the GapCursor appears, the actual result is no cursor movement.

### ProseMirror version

Latest, Gap Cursor 1.0.2

### Affected platforms

Presumably all, tested on **Chrome**.

**Comment by wes-r** [comment]
I don't understand this code well enough to feel confident about this change, but this is what got it working for me:

```diff
--- a/gapcursor.js
+++ b/gapcursor.js
@@ -47,9 +47,9 @@ export class GapCursor extends Selection {
     // Scan up from this position
     for (let d = $pos.depth;; d--) {
       let parent = $pos.node(d)
+      if (!parent) return null
       if (dir > 0 ? $pos.indexAfter(d) < parent.childCount : $pos.index(d) > 0) {
         next = parent.maybeChild(dir > 0 ? $pos.indexAfter(d) : $pos.index(d) - 1)
-        break
       } else if (d == 0) {
         return null
       }
```

**Comment by marijnh** [comment]
The main test case for this code has been the standard horizontal rule nodes, which have `isLeaf = true`. If I try the scenario you describe with such a node, the gap cursor appears to work as intended. Could you be more concrete about the circumstances in which it is breaking?

**Comment by wes-r** [comment]
Ah sorry the bug is really for `selectable: false`. I just added that to the default `horizontal_rule` and the GapCursor stopped working.

**Comment by jordoh** [comment]
Hi @marijnh - the update to prosemirror-gapcursor to address this has introduced an issue when the document contains an `atom: true` node with children, but only if it's in a specific position.

I've created a [glitch example here](https://glitch.com/~reflective-muddy-stetson) using a contrived "container" node:
```
    container: {
      atom: true,
      content: "bullet_list*",
      group: "block",
      toDOM(_node) {
        return [ "div", { class: "container" }, 0 ];
      }
    }
```
To reproduce:
1. Place your cursor in the middle of "line one".
2. Press down arrow on keyboard
Observe that the cursor jumps to the end of the document, as a gap cursor.

I've confirmed that reverting this change results in the expected behavior (the cursor moving down to "line two"), and this happens in both Chrome and Firefox. It doesn't happen if there's a paragraph either before or after the container.

Let me know if I should open a dedicated issue.
